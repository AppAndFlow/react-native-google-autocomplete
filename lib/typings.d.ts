declare type ElementEventTemplate<E> = {
  target: E;
} & Event;

declare type InputEvent = ElementEventTemplate<HTMLInputElement>;
