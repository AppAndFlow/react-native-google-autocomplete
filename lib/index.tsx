export { default as GoogleAutoComplete } from './GoogleAutocomplete';
