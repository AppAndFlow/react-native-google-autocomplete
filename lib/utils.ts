export const isFunction = <T extends Function>(value: any): value is T =>
  typeof value === 'function';
