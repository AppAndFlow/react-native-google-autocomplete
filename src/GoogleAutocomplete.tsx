import { useEffect, useState } from 'react';
import {
  GoogleService,
  type GoogleLocationResult,
} from './services/google.service';
import { useDebounce } from 'use-debounce';
import { useIsMounted } from './useIsMounted';
import { Platform } from 'react-native';

const isWeb = Platform.OS === 'web';

interface Options {
  /**
   * Minimun length of the input before start fetching - default: 2
   */
  minLength?: number;

  /**
   * Debounce request time in ms - default: 300
   */
  debounce?: number;

  /**
   * Language for Google query - default: en
   */
  language?: string;

  /**
   * A grouping of places to which you would like to restrict your results
   */
  components?: string;

  /**
   * See https://developers.google.com/places/web-service/autocomplete#place_types = default: address
   */
  queryTypes?:
    | 'address'
    | 'geocode'
    | '(cities)'
    | 'establishment'
    | 'geocode|establishment';

  /**
   * The distance (in meters) within which to return place results.
   * Note that setting a radius biases results to the indicated area,
   * but may not fully restrict results to the specified area.
   */
  radius?: string;

  /**
   * The latitude to retrieve place information
   */
  lat?: number;

  /**
   * The longitude to retrieve place information
   */
  lng?: number;

  /**
   * Enable strict mode to return search result only in the area defined by radius, lat and lng
   */
  strictBounds?: boolean;

  /**
   * Proxy url if you want to use the web, this is needed cause of CORS issue
   */
  proxyUrl?: string;
}

export const useGoogleAutocomplete = (apiKey: string, opts: Options = {}) => {
  const {
    minLength = 2,
    debounce = 300,
    language = 'en',
    queryTypes = 'address',
  } = opts;
  const isMounted = useIsMounted();
  const [isSearching, setIsSearching] = useState(false);
  const [term, setTerm] = useState('');
  const [debouncedTerm] = useDebounce(term, debounce);
  const [locationResults, setLocationResults] = useState<
    GoogleLocationResult[]
  >([]);
  const [searchError, setSearchError] = useState<Error | null>(null);

  const search = async (value: string) => {
    if (isWeb && !opts.proxyUrl) {
      throw new Error('A proxy url is needed for web');
    }

    setIsSearching(true);
    try {
      const results = await GoogleService.search(
        value,
        {
          key: apiKey,
          language,
          types: queryTypes,
          strictBounds: opts.strictBounds,
          lat: opts.lat,
          lng: opts.lng,
          radius: opts.radius,
        },
        opts.proxyUrl
      );

      setLocationResults(results.predictions);
    } catch (error) {
      if (error instanceof Error) {
        setSearchError(error);
      }
    } finally {
      setIsSearching(false);
    }
  };

  const searchDetails = async (placeId: string) => {
    return GoogleService.searchDetails(placeId, {
      key: apiKey,
      language,
      types: queryTypes,
      components: opts.components,
    });
  };

  const clearSearch = () => {
    if (isMounted()) {
      setLocationResults([]);
      setIsSearching(false);
    }
  };

  useEffect(() => {
    if (debouncedTerm.length >= minLength) {
      search(debouncedTerm);
    }
  }, [debouncedTerm]);

  return {
    locationResults,
    isSearching,
    searchError,
    clearSearch,
    setTerm,
    term,
    searchDetails,
  };
};
