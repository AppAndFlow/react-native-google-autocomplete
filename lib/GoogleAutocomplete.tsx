import * as React from 'react';
import * as PropTypes from 'prop-types';
import { debounce } from 'lodash';

import {
  GoogleLocationResult,
  GoogleService,
  GoogleLocationDetailResult,
} from './services/Google.service';
import { isFunction } from './utils';

export const initialState = {
  inputValue: '',
  locationResults: [],
  isSearching: false,
};

const defaultProps: DefaultProps = {
  /**
   * Minimun length of the input before start fetching - default: 2
   */
  minLength: 2,
  /**
   * Debounce request time in ms - default: 300
   */
  debounce: 300,
  /**
   * Language for Google query - default: en
   */
  language: 'en',

  /**
   * See https://developers.google.com/places/web-service/autocomplete#place_types = default: address
   */
  queryTypes: 'address',
};

export interface DefaultProps {
  minLength: number;
  debounce: number;
  language: string;
  queryTypes:
    | 'address'
    | 'geocode'
    | 'cities'
    | 'establishment'
    | 'geocode|establishment';
}

export type P = Partial<
  {
    children: RenderCallback;
    render: RenderCallback;
    components: string;
    radius: string;
    lat: number;
    lng: number;
  } & DefaultProps
> & {
  /**
   * Api Key provided by google
   */
  apiKey: string;
};

export type RenderCallback = (args: GoogleAutoCompleteProps) => JSX.Element;

export type GoogleAutoCompleteProps = {
  inputValue: S['inputValue'];
  locationResults: S['locationResults'];
  handleTextChange: (value: string) => void;
  handleEventChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  fetchDetails: (placeId: string) => Promise<GoogleLocationDetailResult>;
  clearSearch: () => void;
  isSearching: boolean;
};

export type S = Readonly<{
  inputValue: string;
  locationResults: GoogleLocationResult[];
  isSearching: boolean;
}>;

export class GoogleAutoComplete extends React.PureComponent<P, S> {
  public static readonly defaultProps: DefaultProps = defaultProps;

  static propTypes = {
    /**
     * Minimun length before sending a search request default 2
     */
    minLength: PropTypes.number,
    /**
     * Time for debouncing the search default 300
     */
    debounce: PropTypes.number,
    /**
     * Your api key
     */
    apiKey: PropTypes.string.isRequired,
    /**
     * The language code, indicating in which language the results should be returned, if possible.
     * Searches are also biased to the selected language;
     * results in the selected language may be given a higher ranking.
     */
    language: PropTypes.string,
    /**
     * A grouping of places to which you would like to restrict your results
     */
    components: PropTypes.string,
    /**
     * The distance (in meters) within which to return place results.
     * Note that setting a radius biases results to the indicated area,
     * but may not fully restrict results to the specified area.
     */
    radius: PropTypes.string,

    /**
     * The latitude to retrieve place information
     */
    lat: PropTypes.number,

    /**
     * The longitude to retrieve place information
     */
    lng: PropTypes.number,
  };

  readonly state: S = initialState;

  /**
   * Keep track if the component isMounted or not
   */
  private _isMounted: boolean;

  /**
   * Search to google automplete service
   */
  private _search = debounce(async (term: string) => {
    if (this._isMounted) {
      this.setState({ isSearching: true });

      const searchOpts = {
        key: this.props.apiKey,
        language: this.props.language!,
        types: this.props.queryTypes!,
        components: this.props.components,
        radius: this.props.radius,
        lat: this.props.lat,
        lng: this.props.lng,
      };

      try {
        const results = await GoogleService._search(term, searchOpts);

        this.setState({
          locationResults: results.predictions,
          isSearching: false,
        });
      } catch (error) {
        throw error;
      }
    }
  }, this.props.debounce);

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    const renderProps = {
      inputValue: this.state.inputValue,
      locationResults: this.state.locationResults,
      handleTextChange: this._handleTextChange,
      handleEventChange: this._handleEventChange,
      fetchDetails: this._searchDetails,
      isSearching: this.state.isSearching,
      clearSearch: this._clearSearch,
    };

    if (this.props.render) {
      return this.props.render(renderProps);
    }

    return isFunction(this.props.children)
      ? this.props.children(renderProps)
      : null;
  }

  private _clearSearch = () => {
    if (this._isMounted) {
      this.setState({
        locationResults: [],
        inputValue: '',
      });
    }
  };

  /**
   * Handle the input change for react-native
   */
  private _handleTextChange = (inputValue: string) => {
    if (this.props.apiKey == null) {
      throw new Error('Api Key is required');
    }

    this.setState({
      inputValue,
    });

    if (inputValue.length >= this.props.minLength!) {
      this._search(inputValue);
    }
  };

  /**
   * Handle the input change for react web
   */
  private _handleEventChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    this.setState({
      inputValue: value,
    });

    if (value.length >= this.props.minLength!) {
      this._search(value);
    }
  };

  /**
   * Handle the search details when provide the place_id
   */
  private _searchDetails = async (
    placeId: string,
  ): Promise<GoogleLocationDetailResult> => {
    const searchOpts = {
      key: this.props.apiKey,
      language: this.props.language!,
      types: this.props.queryTypes!,
      components: this.props.components,
    };

    try {
      return GoogleService._searchDetails(placeId, searchOpts);
    } catch (error) {
      throw error;
    }
  };
}
