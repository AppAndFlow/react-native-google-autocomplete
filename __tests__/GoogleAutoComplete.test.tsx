import { View, TextInput, Text } from 'react-native';
import React from 'react';
import { shallow, mount, render } from 'enzyme';

import { GoogleAutoComplete } from '../lib/GoogleAutocomplete';

describe('<GoogleAutoComplete />', () => {
  test('should render correctly', () => {
    const wrapper = shallow(
      <GoogleAutoComplete apiKey="myKey">
        {({ inputValue, handleTextChange }) => (
          <View>
            <TextInput value={inputValue} onChangeText={handleTextChange} />
          </View>
        )}
      </GoogleAutoComplete>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('should return the value from onChangeText to inputValue', () => {
    const wrapper = shallow(
      <GoogleAutoComplete apiKey="myKey">
        {({ inputValue, handleTextChange }) => (
          <TextInput value={inputValue} onChangeText={handleTextChange} />
        )}
      </GoogleAutoComplete>,
    );

    wrapper.find('TextInput').simulate('changeText', '560 av');

    expect(wrapper.find('TextInput').prop('value')).toBe('560 av');
  });

  test('should have props pass and defaultProps', () => {
    const wrapper = mount(
      <GoogleAutoComplete apiKey="myKey">
        {({ inputValue, handleTextChange }) => (
          <View>
            <TextInput value={inputValue} onChangeText={handleTextChange} />
          </View>
        )}
      </GoogleAutoComplete>,
    );

    expect(wrapper.prop('language')).toBe('en');
    expect(wrapper.prop('apiKey')).toBe('myKey');
    expect(wrapper.prop('debounce')).toBe(300);
    expect(wrapper.prop('queryTypes')).toBe('address');
    expect(wrapper.prop('minLength')).toBe(2);
  });

  test('should have the state of the instance equal what the input change', () => {
    const wrapper = shallow(
      <GoogleAutoComplete apiKey="myKey">
        {({ inputValue, handleTextChange }) => (
          <View>
            <TextInput value={inputValue} onChangeText={handleTextChange} />
          </View>
        )}
      </GoogleAutoComplete>,
    );

    wrapper.find('TextInput').simulate('changeText', '560 av');

    const instance = wrapper.instance();

    expect(instance.state.inputValue).toBe('560 av');
  });
});
