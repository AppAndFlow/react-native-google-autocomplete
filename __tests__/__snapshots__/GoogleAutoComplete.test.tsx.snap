// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoogleAutoComplete /> should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GoogleAutoComplete
    apiKey="myKey"
    debounce={300}
    language="en"
    minLength={2}
    queryTypes="address"
  >
    [Function]
  </GoogleAutoComplete>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <TextInput
        onChangeText={[Function]}
        value=""
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "onChangeText": [Function],
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <TextInput
          onChangeText={[Function]}
          value=""
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onChangeText": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
